### PR TITLE
CellMethods now printed in CF-compliant manner

### LINKED ISSUE #3592: Correct CellMethod.__str__ method to be CF-compliant
As shown in the CF-standard (http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/build/ch07s03.html) the correct format for writing a cell method is "coord: method". Iris currently prints "method: coord" instead which should be corrected.

### PR DESCRIPTION
**Pull request by MoseleyS**
### Description

Fixes #3592 

This simple PR rearranges the order in which CellMethod information is formatted by the __str__ method. The grouping of this information (where multiple coords refer to the same method) has therefore been removed.
---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

**Comment by trexfeathers** [comment]
Not being a CF expert, I'm not certain on the specifics here. However I am certain that it breaks the assumptions of the latest cube summary code, having just worked on that (#4206), so some of the doctests will likely fail at the very least.

Have a look at this commit for details. Perhaps you'll have a good idea how to adapt for CF-compliance: 4d8859ccb0c9d20c365695bdbd476cd9f4954fb9 ([specifically the changes in `cube_summary.py`](https://github.com/SciTools/iris/commit/4d8859ccb0c9d20c365695bdbd476cd9f4954fb9#diff-1f7299530bf700aea69dd045391754d224fd4718a4f0a0c3a28d1786ec0af524)).

**Comment by trexfeathers** [comment]
@MoseleyS you'll also need to be working from a **much** more up to date branch, since `MoseleyS:3592_cellmethods_cfcompliant` contains none of the latest CI checks. You can merge `main` into this branch, or rebase it.

**Comment by MoseleyS** [comment]
Looks like I forgot to update my fork first...

**Comment by MoseleyS** [comment]
This should work with the cube_summary changes as the format is changing from `"{method}: {coord}"` to `"{coord}: {method}"` (see #3592). I expect I will need to modify a few more unit tests though to accept the corrected ordering of information. 

The block @trexfeathers pointed me to contains an oddity: `content` looks like it will always be identical to `str(method)` and ought to be refactored as such.

**Comment by trexfeathers** [comment]
@MoseleyS just to note: by all means don't treat the current setup as a straight jacket.

E.g: it may be that cube summary shouldn't be reading `CellMethod`'s string representation at all, but rather working with specialised attributes.

Whatever seems best practice to you, in light of what would be truly CF-compliant.

File: lib/iris/coords.py [hunk 1]
```diff
@@ -2816,25 +2816,24 @@ def __init__(self, method, coords=None, intervals=None, comments=None):
         self._init(method, tuple(_coords), tuple(_intervals), tuple(_comments))
 
     def __str__(self):
```
Comment (trexfeathers) [inline lib/iris/coords.py:2818]: As mentioned in my `cube_summary.py` comment: could you abstract out the code that generates the method + interval + comment string (e.g. `mean (6 minutes, This is a test comment, 12 minutes)`) into its own method: `self.method_str`?

Then just have `self.__str__` stitch together `self.coord_names` and `self.method_str()`? That would avoid the need for downstream string parsing, which can be risky and difficult to follow.

File: lib/iris/_representation/cube_printout.py [hunk 2]
```diff
@@ -257,9 +257,12 @@ def add_scalar_row(name, value=""):
                         add_scalar_row(item.name, item.content)
                         if item.extra:
                             add_scalar_row(item_to_extra_indent + item.extra)
-                elif "attribute" in title or "cell method" in title:
+                elif "attribute" in title:
                     for title, value in zip(sect.names, sect.values):
                         add_scalar_row(title, value)
+                elif "cell method" in title:
+                    for title, value in zip(sect.names, sect.values):
+                        add_scalar_row(", ".join(title), value)
```
Comment (trexfeathers) [inline lib/iris/_representation/cube_printout.py:265]:
```suggestion
                    for coords, method in zip(sect.names, sect.values):
                        add_scalar_row(", ".join(coords), method)
```

Specialised behaviour will benefit from specialised language to help understanding of what's going on.

File: lib/iris/cube.py [hunk 3]
```diff
@@ -750,9 +750,9 @@ class Cube(CFVariableMixin):
                 time                        \
 1998-12-01 00:00:00, bound=(1994-12-01 00:00:00, 1998-12-01 00:00:00)
             Cell methods:
-                mean within years           time
-                mean over years             time
-            Attributes:
+                time                        mean within years
+                time                        mean over years
+             Attributes:
```
Comment (trexfeathers) [inline lib/iris/cube.py:755]:
```suggestion
            Attributes:
```

**Review by trexfeathers — CHANGES_REQUESTED** [review]
Thanks for slogging through this @MoseleyS, I think it's a great idea! I'd also appreciate an opinion from @pp-mo about the general concept here, since @pp-mo has more experience with CF.

I have a few comments for you. The main issue I have is the downstream parsing of `CellMethod.__str__` in the cube summary code. I already struggled to de-risk this in my previous implementation, and now with your improved implementation that issue has become that extra bit problematic. See what you think of my suggestion.

Once all other changes are made, you will need to find existing examples in the documentation and change them to the new world before the doctests will pass. If it helps, all the failures are listed in the CI details ([here for latest run at time of writing](https://cirrus-ci.com/task/5681042687262720)).

**Comment by MoseleyS** [comment]
Thanks for the suggestions. I'm a bit busy with other work, so this is going to have to wait for a week or so, but in principle, I agree with your suggestions.

**Comment by trexfeathers** [comment]
@MoseleyS does this still matter to you? Can we get it over the line? Or should I close it?

Could perhaps bring to a UKMO 'Surgery'?

**Comment by MoseleyS** [comment]
> @MoseleyS does this still matter to you? Can we get it over the line? Or should I close it?
> 
> Could perhaps bring to a UKMO 'Surgery'?

I'd forgotten about this. I'd still like it to happen as I like consistency. I'm not sure when I'm going to have time to look at this again though, perhaps next month. Feel free to take over if you'd like to.

**Comment by trexfeathers** [comment]
Thanks @MoseleyS. @pp-mo has pointed out there are arguments for the current implementation too, although we're not certain in either direction. I'll mark this for team discussion just to make sure everyone is happy.

**Comment by trexfeathers** [comment]
Summary from team discussion today (ping @pp-mo):

We still agree that the `CellMethod.__str__()` should print something CF compliant.

More complex for a `Cube`. The left hand text uses the names of objects wherever possible. This dimension is called `foo`, this coordinate is called `bar`, etcetera. That is where the 'controversy' lies - there is no clear answer for cell methods, since they do not have names but are instead provided as a sequence of operations that were performed on this data. So we'd like to go with a THIRD WAY 😁 :

```
    Cell methods:
        0                < my_cube.cell_methods[0].__str__() >
        1                < my_cube.cell_methods[1].__str__() >
        2                < my_cube.cell_methods[2].__str__() >
```

I still intend to pick this up, but likely not before December. @MoseleyS is free to pursue before then if they wish.

**Comment by MoseleyS** [comment]
> ```
>     Cell methods:
>         0                < my_cube.cell_methods[0].__str__() >
>         1                < my_cube.cell_methods[1].__str__() >
>         2                < my_cube.cell_methods[2].__str__() >
> ```
> 
> I still intend to pick this up, but likely not before December. @MoseleyS is free to pursue before then if they wish.

That's fine by me. Ideally, the `CellMethod.__str__` method will return the CF-compliant format.
I doubt I'll get to this before December, but anything is possible I guess.

**Comment by trexfeathers** [comment]
Replaced by #5224

**Pull request closed without merging**